File: mockserver-ui/src/lib/llmPricing.ts
/**
 * Estimated cost calculation for LLM API usage.
 *
 * Pure function that consults a pricing table keyed on (provider, model)
 * with inputPerMillion and outputPerMillion rates in USD.
 *
 * Returns null for unknown models so the UI can display "—".
 *
 * Pricing source / freshness
 * --------------------------
 * Rates below are public provider list prices captured 2026-06 (newer
 * families) / 2025-Q4 (original entries). They WILL drift; treat the
 * dashboard total as an estimate, not an invoice.
 * Refresh procedure: update the relevant `*_PRICING` array entry and
 * append the source URL + capture date to its inline comment.
 *
 * IMPORTANT — companion source of truth:
 * This table is a parallel copy of the Java pricing table at
 * `mockserver/mockserver-core/src/main/java/org/mockserver/llm/cost/LlmPricing.java`,
 * which is the source of truth. When refreshing rates or adding/removing
 * models, UPDATE BOTH `llmPricing.ts` AND `LlmPricing.java` TOGETHER so the
 * dashboard estimate matches the server-side cost figures. The drift guard in
 * `llmPricing.test.ts` encodes the canonical expected prices and will fail if
 * this table drifts from them.
 *
 * Sources:
 *   - Anthropic: https://www.anthropic.com/pricing
 *   - OpenAI:    https://openai.com/api/pricing
 *   - Gemini:    https://ai.google.dev/pricing
 */

// ---------------------------------------------------------------------------
// Pricing table
// ---------------------------------------------------------------------------

interface PricingEntry {
  inputPerMillion: number;
  outputPerMillion: number;
}

/**
 * Pricing entries keyed by a model-name prefix. The lookup walks this list
 * in order and uses the first entry whose key is a prefix of the model id
 * (case-insensitive). More-specific prefixes must appear before less-specific
 * ones so that e.g. "claude-opus-4-8" matches the 5/25 entry before the
 * generic "claude-opus-4" 15/75 catch-all, and "gpt-4o-mini" before "gpt-4o".
 *
 * Mirrors `LlmPricing.java` (the source of truth) — keep both in sync.
 */
const ANTHROPIC_PRICING: Array<[string, PricingEntry]> = [
  // Current Claude families (public list prices, anthropic.com/pricing,
  // captured 2026-06). More-specific dotted prefixes (claude-opus-4-8) come
  // before the generic claude-opus-4 so they win the prefix walk.
  ['claude-fable-5', { inputPerMillion: 10.0, outputPerMillion: 50.0 }],
  ['claude-opus-4-8', { inputPerMillion: 5.0, outputPerMillion: 25.0 }],
  ['claude-opus-4-7', { inputPerMillion: 5.0, outputPerMillion: 25.0 }],
  ['claude-opus-4-6', { inputPerMillion: 5.0, outputPerMillion: 25.0 }],
  ['claude-opus-4-5', { inputPerMillion: 5.0, outputPerMillion: 25.0 }],
  ['claude-sonnet-4-6', { inputPerMillion: 3.0, outputPerMillion: 15.0 }],
  ['claude-sonnet-4-5', { inputPerMillion: 3.0, outputPerMillion: 15.0 }],
  ['claude-haiku-4-5', { inputPerMillion: 1.0, outputPerMillion: 5.0 }],
  // Original Claude 4.0/4.1 families (claude-opus-4-0/4-1, claude-sonnet-4-0).
  ['claude-opus-4', { inputPerMillion: 15.0, outputPerMillion: 75.0 }],
  ['claude-sonnet-4', { inputPerMillion: 3.0, outputPerMillion: 15.0 }],
  ['claude-haiku-4', { inputPerMillion: 0.8, outputPerMillion: 4.0 }],
];

const OPENAI_PRICING: Array<[string, PricingEntry]> = [
  // ORDERING MATTERS — the lookup uses startsWith, so the longer
  // `gpt-4o-mini` MUST appear before `gpt-4o`. Reversing the order would
  // cause every `gpt-4o-mini` model to silently bill at gpt-4o's rate.
  ['gpt-4o-mini', { inputPerMillion: 0.15, outputPerMillion: 0.6 }],
  ['gpt-4o', { inputPerMillion: 2.5, outputPerMillion: 10.0 }],
  // gpt-4.1 family: list prices captured 2026-06 (openai.com/api/pricing).
  ['gpt-4.1-mini', { inputPerMillion: 0.4, outputPerMillion: 1.6 }],
  ['gpt-4.1-nano', { inputPerMillion: 0.1, outputPerMillion: 0.4 }],
  ['gpt-4.1', { inputPerMillion: 2.0, outputPerMillion: 8.0 }],
  // o-series reasoning models. o4-mini / o3-mini before o3 in the walk.
  ['o4-mini', { inputPerMillion: 1.1, outputPerMillion: 4.4 }],
  ['o3-mini', { inputPerMillion: 1.1, outputPerMillion: 4.4 }],
  ['o3', { inputPerMillion: 15.0, outputPerMillion: 60.0 }],
  // APPROXIMATE: gpt-5* prices are NOT verified — mapped to the nearest known
  // tier (gpt-4o flagship / gpt-4o-mini cheap tier) as a placeholder so a
  // recognised model resolves to SOMETHING rather than null. Confirm against
  // the provider price list before relying on these figures.
  ['gpt-5-mini', { inputPerMillion: 0.15, outputPerMillion: 0.6 }], // ~gpt-4o-mini tier (approx, TBC)
  ['gpt-5-nano', { inputPerMillion: 0.15, outputPerMillion: 0.6 }], // ~gpt-4o-mini tier (approx, TBC)
  ['gpt-5', { inputPerMillion: 2.5, outputPerMillion: 10.0 }], // ~gpt-4o flagship tier (approx, TBC)
];

const GEMINI_PRICING: Array<[string, PricingEntry]> = [
  // gemini-2.5 families: list prices captured 2026-06 (ai.google.dev/pricing).
  // More-specific 2.5-flash-lite before 2.5-flash before 2.5-pro in the walk.
  ['gemini-2.5-flash-lite', { inputPerMillion: 0.1, outputPerMillion: 0.4 }],
  ['gemini-2.5-flash', { inputPerMillion: 0.3, outputPerMillion: 2.5 }],
  ['gemini-2.5-pro', { inputPerMillion: 1.25, outputPerMillion: 10.0 }],
  ['gemini-2.0-flash-lite', { inputPerMillion: 0.075, outputPerMillion: 0.3 }],
  ['gemini-2.0-flash', { inputPerMillion: 0.1, outputPerMillion: 0.4 }],
];

// Ollama is always free (local models)
const OLLAMA_PRICING: PricingEntry = { inputPerMillion: 0, outputPerMillion: 0 };

// ---------------------------------------------------------------------------
// Lookup helpers
// ---------------------------------------------------------------------------

function findPricing(table: Array<[string, PricingEntry]>, model: string): PricingEntry | null {
  const lower = model.toLowerCase();
  for (const [prefix, entry] of table) {
    if (lower.startsWith(prefix.toLowerCase())) {
      return entry;
    }
  }
  return null;
}

/**
 * Look up pricing for a given provider and model.
 * Returns null if the model is not recognised.
 */
function getPricing(provider: string, model: string): PricingEntry | null {
  const kind = provider.toLowerCase();

  if (kind === 'anthropic' || kind === 'bedrock') {
    // parseTraffic classifies Bedrock Claude traffic as kind 'anthropic', and its model
    // ids are prefixed (e.g. "anthropic.claude-sonnet-4-…" or, with inference profiles,
    // "us.anthropic.claude-…"). Strip up to and including "anthropic." so the claude-*
    // pricing keys match; a bare "claude-…" id is unaffected.
    const stripped = model.replace(/^.*anthropic\./, '');
    return findPricing(ANTHROPIC_PRICING, stripped);
  }
  if (kind === 'openai' || kind === 'azure_openai') {
    // Azure OpenAI traffic uses *deployment names* (user-defined, e.g.
    // "my-gpt4o-prod") rather than canonical OpenAI model ids, so this
    // lookup will return null for most real Azure deployments. The UI
    // surfaces this as "—" — document that limitation if you wire an
    // Azure-detection path into `parseTraffic`.
    return findPricing(OPENAI_PRICING, model);
  }
  if (kind === 'openai_responses') {
    return findPricing(OPENAI_PRICING, model);
  }
  if (kind === 'gemini') {
    return findPricing(GEMINI_PRICING, model);
  }
  if (kind === 'ollama') {
    return OLLAMA_PRICING;
  }

  return null;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Estimate the cost in USD for a given provider, model, and token counts.
 *
 * @returns The estimated cost in USD, or `null` if the model is unknown.
 *          Returns `0` for zero tokens even on a known model.
 */
export function estimateCostUsd(
  provider: string,
  model: string,
  inputTokens: number,
  outputTokens: number,
): number | null {
  const pricing = getPricing(provider, model);
  if (pricing === null) return null;

  if (inputTokens === 0 && outputTokens === 0) return 0;

  return (
    (inputTokens / 1_000_000) * pricing.inputPerMillion +
    (outputTokens / 1_000_000) * pricing.outputPerMillion
  );
}

// ---------------------------------------------------------------------------
// Drift-guard support
// ---------------------------------------------------------------------------

/**
 * The ordered model-id prefixes present in each provider's pricing table.
 * Exported solely so the drift-guard test (`llmPricing.test.ts`) can assert
 * this UI table stays a superset of the canonical model list mirrored from
 * `LlmPricing.java`. Not part of the runtime UI surface.
 */
export const PRICING_MODEL_PREFIXES: Record<string, string[]> = {
  anthropic: ANTHROPIC_PRICING.map(([prefix]) => prefix),
  openai: OPENAI_PRICING.map(([prefix]) => prefix),
  gemini: GEMINI_PRICING.map(([prefix]) => prefix),
};

File: mockserver/mockserver-core/src/main/java/org/mockserver/llm/cost/LlmPricing.java
package org.mockserver.llm.cost;

import org.mockserver.model.Provider;

import java.util.AbstractMap.SimpleImmutableEntry;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * Estimated cost calculation for LLM API usage.
 *
 * <p>A pure lookup over a pricing table keyed on (provider, model) with
 * input/output rates in USD per million tokens. Returns {@code null} for an
 * unknown model so callers can distinguish "free/known-zero" from "unpriceable".
 *
 * <p><strong>Companion dashboard pricing table</strong>
 * {@code mockserver-ui/src/lib/llmPricing.ts} maintains a parallel table with
 * the same prefix-walk semantics, and is kept in sync with this table. A
 * UI-side drift guard ({@code mockserver-ui/src/__tests__/llmPricing.test.ts})
 * snapshots the expected prefixes and rates and fails if the dashboard table
 * drifts from them, so when refreshing rates update both tables together and
 * update that test's canonical snapshot to match.
 *
 * <p><strong>Pricing source / freshness:</strong> rates are public provider
 * list prices captured 2026-06 (newer families) / 2025-Q4 (original entries).
 * They WILL drift — treat any total as an estimate, not an invoice. Sources:
 * Anthropic (anthropic.com/pricing), OpenAI (openai.com/api/pricing),
 * Gemini (ai.google.dev/pricing).
 */
public final class LlmPricing {

    /**
     * The cheapest model in a provider's pricing table — its prefix label and
     * blended (input+output per-million) rate. Returned by {@link #cheapestModel(Provider)}
     * so a caller can suggest a smaller model for trivial work.
     */
    public static final class ModelOption {
        private final String label;
        private final double blendedPerMillion;

        public ModelOption(String label, double blendedPerMillion) {
            this.label = label;
            this.blendedPerMillion = blendedPerMillion;
        }

        public String getLabel() {
            return label;
        }

        public double getBlendedPerMillion() {
            return blendedPerMillion;
        }
    }

    /** Input/output list price in USD per million tokens. */
    public static final class PricingEntry {
        private final double inputPerMillion;
        private final double outputPerMillion;

        public PricingEntry(double inputPerMillion, double outputPerMillion) {
            this.inputPerMillion = inputPerMillion;
            this.outputPerMillion = outputPerMillion;
        }

        public double getInputPerMillion() {
            return inputPerMillion;
        }

        public double getOutputPerMillion() {
            return outputPerMillion;
        }
    }

    // Prefix-keyed tables. The lookup walks each list in order and uses the
    // first entry whose key is a prefix of the model id (case-insensitive), so
    // MORE-SPECIFIC PREFIXES MUST COME FIRST (e.g. gpt-4o-mini before gpt-4o,
    // otherwise every gpt-4o-mini model would silently bill at gpt-4o's rate).
    private static final List<Map.Entry<String, PricingEntry>> ANTHROPIC_PRICING = Arrays.asList(
        // Current Claude families (public list prices, anthropic.com/pricing, captured
        // 2026-06). More-specific dotted prefixes (claude-opus-4-8) come before the
        // generic claude-opus-4 so they win the prefix walk.
        entry("claude-fable-5", new PricingEntry(10.0, 50.0)),
        entry("claude-opus-4-8", new PricingEntry(5.0, 25.0)),
        entry("claude-opus-4-7", new PricingEntry(5.0, 25.0)),
        entry("claude-opus-4-6", new PricingEntry(5.0, 25.0)),
        entry("claude-opus-4-5", new PricingEntry(5.0, 25.0)),
        entry("claude-sonnet-4-6", new PricingEntry(3.0, 15.0)),
        entry("claude-sonnet-4-5", new PricingEntry(3.0, 15.0)),
        entry("claude-haiku-4-5", new PricingEntry(1.0, 5.0)),
        // Original Claude 4.0/4.1 families (claude-opus-4-0/4-1, claude-sonnet-4-0).
        entry("claude-opus-4", new PricingEntry(15.0, 75.0)),
        entry("claude-sonnet-4", new PricingEntry(3.0, 15.0)),
        entry("claude-haiku-4", new PricingEntry(0.8, 4.0))
    );

    private static final List<Map.Entry<String, PricingEntry>> OPENAI_PRICING = Arrays.asList(
        // gpt-4o-mini before gpt-4o (prefix walk). gpt-4.1 family: list prices
        // captured 2026-06 (openai.com/api/pricing).
        entry("gpt-4o-mini", new PricingEntry(0.15, 0.6)),
        entry("gpt-4o", new PricingEntry(2.5, 10.0)),
        entry("gpt-4.1-mini", new PricingEntry(0.4, 1.6)),
        entry("gpt-4.1-nano", new PricingEntry(0.1, 0.4)),
        entry("gpt-4.1", new PricingEntry(2.0, 8.0)),
        // o-series reasoning models. o4-mini / o3-mini before o3 in the walk.
        entry("o4-mini", new PricingEntry(1.1, 4.4)),
        entry("o3-mini", new PricingEntry(1.1, 4.4)),
        entry("o3", new PricingEntry(15.0, 60.0)),
        // APPROXIMATE: gpt-5* prices are NOT verified here — mapped to the nearest
        // known tier (gpt-4o flagship / gpt-4o-mini cheap tier) as a placeholder so a
        // recognised model resolves to SOMETHING rather than null. Confirm against the
        // provider price list before relying on these figures.
        entry("gpt-5-mini", new PricingEntry(0.15, 0.6)),   // ~gpt-4o-mini tier (approx, TBC)
        entry("gpt-5-nano", new PricingEntry(0.15, 0.6)),   // ~gpt-4o-mini tier (approx, TBC)
        entry("gpt-5", new PricingEntry(2.5, 10.0))         // ~gpt-4o flagship tier (approx, TBC)
    );

    private static final List<Map.Entry<String, PricingEntry>> GEMINI_PRICING = Arrays.asList(
        // gemini-2.5 families: list prices captured 2026-06 (ai.google.dev/pricing).
        // More-specific 2.5-flash-lite before 2.5-flash before 2.5-pro in the walk.
        entry("gemini-2.5-flash-lite", new PricingEntry(0.1, 0.4)),
        entry("gemini-2.5-flash", new PricingEntry(0.3, 2.5)),
        entry("gemini-2.5-pro", new PricingEntry(1.25, 10.0)),
        entry("gemini-2.0-flash-lite", new PricingEntry(0.075, 0.3)),
        entry("gemini-2.0-flash", new PricingEntry(0.1, 0.4))
    );

    // Ollama is always free (local models).
    private static final PricingEntry OLLAMA_PRICING = new PricingEntry(0.0, 0.0);

    private LlmPricing() {
    }

    private static Map.Entry<String, PricingEntry> entry(String prefix, PricingEntry pricing) {
        return new SimpleImmutableEntry<>(prefix, pricing);
    }

    private static PricingEntry findPricing(List<Map.Entry<String, PricingEntry>> table, String model) {
        if (model == null) {
            return null;
        }
        String lower = model.toLowerCase();
        for (Map.Entry<String, PricingEntry> e : table) {
            if (lower.startsWith(e.getKey().toLowerCase())) {
                return e.getValue();
            }
        }
        return null;
    }

    /**
     * Look up pricing for a provider and model, or {@code null} if unrecognised.
     */
    public static PricingEntry getPricing(Provider provider, String model) {
        if (provider == null) {
            return null;
        }
        switch (provider) {
            case ANTHROPIC:
                return findPricing(ANTHROPIC_PRICING, model);
            case BEDROCK:
                // Bedrock Anthropic model ids look like "anthropic.claude-sonnet-4-…".
                return findPricing(ANTHROPIC_PRICING, model == null ? null : model.replaceFirst("^anthropic\\.", ""));
            case OPENAI:
            case AZURE_OPENAI:
                // Azure OpenAI uses user-defined deployment names rather than canonical
                // model ids, so this returns null for most real Azure deployments.
                return findPricing(OPENAI_PRICING, model);
            case OPENAI_RESPONSES:
                return findPricing(OPENAI_PRICING, model);
            case GEMINI:
                return findPricing(GEMINI_PRICING, model);
            case OLLAMA:
                return OLLAMA_PRICING;
            default:
                return null;
        }
    }

    /**
     * Estimate the cost in USD for a provider, model, and token counts.
     *
     * @return the estimated cost in USD, or {@code null} if the model is
     * unknown. Returns {@code 0.0} for zero tokens on a known (priced) model.
     */
    public static Double estimateCostUsd(Provider provider, String model, long inputTokens, long outputTokens) {
        PricingEntry pricing = getPricing(provider, model);
        if (pricing == null) {
            return null;
        }
        if (inputTokens == 0 && outputTokens == 0) {
            return 0.0;
        }
        return (inputTokens / 1_000_000.0) * pricing.getInputPerMillion()
            + (outputTokens / 1_000_000.0) * pricing.getOutputPerMillion();
    }

    /**
     * Return the provider's pricing table, or {@code null} when the provider has
     * no prefix table (OLLAMA flat-free, AZURE deployment names, unknown).
     */
    private static List<Map.Entry<String, PricingEntry>> tableFor(Provider provider) {
        if (provider == null) {
            return null;
        }
        switch (provider) {
            case ANTHROPIC:
            case BEDROCK:
                return ANTHROPIC_PRICING;
            case OPENAI:
            case OPENAI_RESPONSES:
                return OPENAI_PRICING;
            case GEMINI:
                return GEMINI_PRICING;
            default:
                // AZURE_OPENAI (deployment names), OLLAMA (flat free), unknown.
                return null;
        }
    }

    /**
     * The cheapest model in the provider's pricing table by blended
     * (input+output per-million) rate, or {@code null} for OLLAMA / unknown /
     * tableless providers. Deterministic: ties resolve to the first table entry.
     */
    public static ModelOption cheapestModel(Provider provider) {
        List<Map.Entry<String, PricingEntry>> table = tableFor(provider);
        if (table == null || table.isEmpty()) {
            return null;
        }
        Map.Entry<String, PricingEntry> cheapest = null;
        double cheapestBlended = Double.MAX_VALUE;
        for (Map.Entry<String, PricingEntry> e : table) {
            double blended = e.getValue().getInputPerMillion() + e.getValue().getOutputPerMillion();
            if (blended < cheapestBlended) {
                cheapestBlended = blended;
                cheapest = e;
            }
        }
        return cheapest == null ? null : new ModelOption(cheapest.getKey(), cheapestBlended);
    }

    /**
     * The blended (input+output per-million) rate for a provider/model, or
     * {@code null} when the model is unpriced.
     */
    public static Double blendedPerMillion(Provider provider, String model) {
        PricingEntry pricing = getPricing(provider, model);
        if (pricing == null) {
            return null;
        }
        return pricing.getInputPerMillion() + pricing.getOutputPerMillion();
    }
}
